src/idl_gen_cpp.cpp

```diff
@@ -2417,8 +2417,20 @@ class CppGenerator : public BaseGenerator {
 
     // Generate KeyCompareWithValue function
     if (is_string) {
+      // Compares key against a null-terminated char array.
       code_ += "  int KeyCompareWithValue(const char *_{{FIELD_NAME}}) const {";
       code_ += "    return strcmp({{FIELD_NAME}}()->c_str(), _{{FIELD_NAME}});";
+      code_ += "  }";
+      // Compares key against any string-like object (e.g. std::string_view or
+      // std::string) that implements operator< comparison with const char*.
+      code_ += "  template<typename StringType>";
+      code_ +=
+          "  int KeyCompareWithValue(const StringType& _{{FIELD_NAME}}) const "
+          "{";
+      code_ +=
+          "    if ({{FIELD_NAME}}()->c_str() < _{{FIELD_NAME}}) return -1;";
```

Comment:
Can we add a function somewhere that just takes two parameters to do the comparison, and then the codegen just uses that? It reduces the codegen output, and then we can also test this new function in isolation.

Reply:
I suppose we can, but the function would have to remain a template (since StringType is unknown) so it'd still have to be inline and instantiated at the call site. There might be cases when the compiler would reuse instantiations but I'm not sure if it's worth the coding effort? Especially since the compiler's optimizer could inline the two-parameter function anyway, and it'd end up becoming the same assembly :)

Reply:
Yeah, i was thinking more for readability than performance. But we can always change this later.